rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,156 @@
+- `date`: Required. An ISO 8601 `yyyy-mm-dd`-formatted date from when the issue was added.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+
+Example values:
+
+```nix
+meta.issues = [{
+  kind = "deprecated";
+  message = "This package depends on Python 2, which has reached end of life.";
+  date = "1970-01-01";
+  urls = [ "https://github.com/NixOS/nixpkgs/issues/148779" ];
+} {
+  kind = "removal";
+  message = "The application has been abandoned upstream, use libfoo instead";
+  date = "1970-01-01";
+}];
+```
+
+### nixpkgs integration
+
+Two new config options are added to nixpkgs, `ignoreWarningsPredicate` and `ignoreWarningsPackages`. A new environment variable is defined, `NIXPKGS_IGNORE_WARNINGS`. Their semantic and implementation directly parallel the existing "insecure" package handling.
+
+Similarly to broken, insecure and unfree packages, evaluating a package with an issue fails evaluation. Ignoring a package without issues (i.e. they have all been resolved) results in a warning at evaluation time.
+
+## Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Package removal
+
+There are two ways issues interact with the removal of packages: Either they get an issue because they are going to be removed, or they are removed because they have an open issue for a prolonged period of time.
+
+- Instead of removing a package directly, it should first get an issue announcing the planned removal. This will allow users to migrate away beforehand. `removal` must be used as `kind` (This will facilitate automation in the future).
+- Before branch-off for a new release, all (leaf) packages with issues that predate the previous branch-off are deemed safe for removal (unless stated otherwise). If a package is removed based on its issue, that message becomes part of the new `throw` alias.
+
+### Propagation across transitive dependencies
+
+When a package that is depended on has an issue, all packages that depend on it will fail to evaluate until that package is ignored or the issue resolved. Sometimes, this is sufficient.
+
+When the issue requires actions on dependents however, it does not sufficiently inform about all packages that need action. Marking all dependents with that issue is not a good idea either though: it would require users to go through some potentially long dependency chains. Instead, only applications, leaf packages or packages with very few dependents should get the issue.
+
+As an example, take `gksu` with the `gksu` → `libgksu` → `libglade` → `python2` dependency chain (for the sake of the example, ignore that it also depends on EOL Gtk 2). Obviously, `python2` should get an issue. As a leaf/application, `gksu` should get one too (it could be the same, or with an adpated message). For the packages in between, it depends on whether they require individual action or not.
```

Comment:
Does this mean that in an evaluation that involves gksu, both the warning for gksu and the warning for python2 will be printed, even if that's the only path through which the evaluation depends on python2? I'm not sure I like that.

Reply:
Since warnings are evaluation errors, the evaluation stops after printing the first warning.

Reply:
That seems painful if there's no way to opt out, if you're adding a new warning type you should be able to get it to eval as far as it can and list everything. Also [Hydra](https://github.com/nixos/nixpkgs/blob/022caabb5f2265ad4006c1fa5b1ebe69fb0c3faf/pkgs/stdenv/generic/check-meta.nix#L190-L194).

Reply:
In the updated proposal most of the time problems will be printed via `builtins.trace`. However, in the cases that fail eval with `throw` it will stop at the first error. There's no easy way around that, and fixing that is out of scope of this RFC.

Reply:
"issues and warnings" in the title, «problems» in your message, «warning»s being used almost exclusively in the RFC text. Needs better terminology.

Reply:
Yeah the terms keep changing because naming things is hard and people remain unsatisfied. If you count the original "advisories" from the pre-RFC, we're in the fourth naming iteration or something.